packages/gatsby/src/commands/develop.ts

```diff
@@ -1,613 +1,200 @@
-    })
-    activity.end()
-    activity = report.activityTimer(`rebuild schema`)
-    activity.start()
-    await rebuildSchema({ parentSpan: activity })
-    activity.end()
-    startSchemaHotReloader()
   }
-  app.use(REFRESH_ENDPOINT, express.json())
-  app.post(REFRESH_ENDPOINT, (req, res) => {
-    const enableRefresh = process.env.ENABLE_GATSBY_REFRESH_ENDPOINT
-    const refreshToken = process.env.GATSBY_REFRESH_TOKEN
-    const authorizedRefresh =
-      !refreshToken || req.headers.authorization === refreshToken
-
-    if (enableRefresh && authorizedRefresh) {
-      refresh(req)
+  async stop(signal: string | null = null, code?: number): Promise<void> {
+    this.isRunning = false
+    if (signal) {
+      this.process.kill(signal)
+    } else {
+      this.process.exit(code)
     }
-    res.end()
-  })
 
-  app.get(`/__open-stack-frame-in-editor`, (req, res) => {
-    launchEditor(req.query.fileName, req.query.lineNumber)
-    res.end()
-  })
-
-  // Disable directory indexing i.e. serving index.html from a directory.
-  // This can lead to serving stale html files during development.
-  //
-  // We serve by default an empty index.html that sets up the dev environment.
-  app.use(developStatic(`public`, { index: false }))
-
-  app.use(
-    webpackDevMiddleware(compiler, {
-      logLevel: `silent`,
-      publicPath: devConfig.output.publicPath,
-      watchOptions: devConfig.devServer
-        ? devConfig.devServer.watchOptions
-        : null,
-      stats: `errors-only`,
-    })
-  )
-
-  // Expose access to app for advanced use cases
-  const { developMiddleware } = store.getState().config
-
-  if (developMiddleware) {
-    developMiddleware(app, program)
-  }
-
-  // Set up API proxy.
-  const { proxy } = store.getState().config
-  if (proxy) {
-    proxy.forEach(({ prefix, url }) => {
-      app.use(`${prefix}/*`, (req, res) => {
-        const proxiedUrl = url + req.originalUrl
-        const {
-          // remove `host` from copied headers
-          // eslint-disable-next-line @typescript-eslint/no-unused-vars
-          headers: { host, ...headers },
-          method,
-        } = req
-        req
-          .pipe(
-            got
-              .stream(proxiedUrl, { headers, method, decompress: false })
-              .on(`response`, response =>
-                res.writeHead(response.statusCode || 200, response.headers)
-              )
-              .on(`error`, (err, _, response) => {
-                if (response) {
-                  res.writeHead(response.statusCode || 400, response.headers)
-                } else {
-                  const message = `Error when trying to proxy request "${req.originalUrl}" to "${proxiedUrl}"`
-
-                  report.error(message, err)
-                  res.sendStatus(500)
-                }
-              })
-          )
-          .pipe(res)
+    return new Promise(resolve => {
+      this.process.on(`exit`, () => {
+        this.process.removeAllListeners()
```

Comment:
Note: I added this because `process.kill` can't be awaited on and is _not_ guaranteed to finish  synchronously in case we have exit handlers. This hack lets us send a kill signal and then listen for the `exit` event from the child process.